File: account-server/server_api/REFRESH_GAMESERVER_INSTANCE_STATE.py
import json
import datetime

# REFRESH_GAMESERVER_INSTANCE_STATE:
# 1. El controller (del gameserver node) envia el lobby_id.
# 2. El servidor actualiza el estado de la instancia en la lista de gameserver instances.
#    Si no existe el lobby_id en la lista de los game servers entonces devuelve lobby_id vacio.
async def REFRESH_GAMESERVER_INSTANCE_STATE(req, ws, send_message, gameservers_info):

    # Aqui se debe comprobar que la req es correcta y en caso contrario se ignora sin enviar nada al cliente.
    if not all(k in req for k in ('request_id', 'lobby_id')):
        print("ERROR: REFRESH_GAMESERVER_INSTANCE_STATE bad input parameters.")
        return

    lobby_id = req['lobby_id']
    gsinstances = gameservers_info['instances']
    if lobby_id in gsinstances:
        entry = gsinstances[lobby_id]
        expire_time = entry['expire_time']
        if expire_time >= datetime.datetime.now():
            entry['expire_time'] = datetime.datetime.now() + gameservers_info['timeout']
        else:
            gameservers_info.pop(lobby_id, None)
    else:
        lobby_id = ""
    
    response = {
        'request_id': req['request_id'],
        'lobby_id': lobby_id
    }

    await send_message(response, ws)   # send response

File: account-server-golang/server_api/REFRESH_GAMESERVER_INSTANCE_STATE.go
package server_api

import (
	"account-server/gameservers"
	"account-server/utils"
	"encoding/json"
	"errors"
)

type request_REFRESH_GAMESERVER_INSTANCE_STATE struct {
	Request_id  string
	Lobby_id    string
	Current_map string
}

//REFRESH_GAMESERVER_INSTANCE_STATE:
// 1. El controller (del gameserver node) envia el lobby_id.
// 2. El servidor actualiza el estado de la instancia en la lista de gameserver instances.
//    Si no existe el lobby_id en la lista de los game servers entonces devuelve lobby_id vacio.
func REFRESH_GAMESERVER_INSTANCE_STATE(message []byte, sendResponse func([]byte)) error {

	var r request_REFRESH_GAMESERVER_INSTANCE_STATE

	// Comprueba que estan todos los parametros requeridos.
	{
		var m map[string]any
		err := json.Unmarshal(message, &m)
		if err != nil {
			return err
		}
		if !utils.ContainsAllKeys(m, []string{"request_id", "lobby_id", "current_map"}) {
			err = errors.New("REFRESH_GAMESERVER_INSTANCE_STATE bad input parameters")
			return err
		}
	}

	err := json.Unmarshal(message, &r)
	if err != nil {
		return err
	}

	lobbyId := r.Lobby_id
	ok := gameservers.RefreshGameServer(lobbyId, r.Current_map)
	if !ok {
		lobbyId = ""
	}

	response, err := json.Marshal(map[string]any{
		"request_id": r.Request_id,
		"lobby_id":   lobbyId,
	})

	if err != nil {
		return err
	}

	// fmt.Println("Enviada respuesta para REFRESH_GAMESERVER_INSTANCE_STATE request_id: ", r.Request_id)
	sendResponse(response)

	return err
}
